### PR TITLE
[Cache] Use namespace versioning for backends that dont support clearing by keys

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.3
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #23960
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Cache/Traits/AbstractTrait.php [hunk 1]
```diff
@@ -189,11 +204,18 @@ private function getId($key)
     {
         CacheItem::validateKey($key);
 
+        if ($this->versioningIsEnabled && '' === $this->namespaceVersion) {
```
Comment (jderusse) [inline src/Symfony/Component/Cache/Traits/AbstractTrait.php]: Other classes used the trick `isset($string[0])` to check if their is a namespace.

**Comment by nicolas-grekas** [comment]
with new false positive from php-cs-fixer...

File: src/Symfony/Component/Cache/Traits/AbstractTrait.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/Cache/Traits/AbstractTrait.php]: not sure it applies here

**Review by hcomnetworkers — APPROVED** [review]
Looks good! 

Although I'm wondering why you put a colon after the namespace while colons are not allowed in validateKey?
(We used to use colons in our cache keys until we switched to the new classes which don't allow it ;) 
Also on a side note: Is there any reason why an integer is not allowed as cache key if it gets namespaced anyway? We have/had that, unfortunately.)

**Comment by nicolas-grekas** [comment]
> put a colon after the namespace while colons are not allowed in validateKey

I put a colon *because* they are not allowed *when give from the outside*. But here, we're *past* the interface, so the PSR-6/16 constraint do not apply. Which is very nice for us: using a colon makes it impossible to have any confusion with user input.

> why an integer is not allowed as cache key

because PSR-6/16 states that a key must be a string. So `6` is not allowed, but `"6"` is.

**Comment by andrerom** [comment]
A bit ton on this, would it be possible to rather do this on init, and using multi get together with the item lookup? Then if version has changed either refetch the times, or once there is stale cache support on get we can still return the items as the calee don't need super fresh items.

**Comment by nicolas-grekas** [comment]
As the PR is now, the version retrieval is done only once, on first need. That's how Doctrine does it also afaik. Isn't it good enough?

**Comment by andrerom** [comment]
Ah, sorry missed that. It's fine with me, but you can bet there will be bug report for this down the line on long running processes :)

**Comment by nicolas-grekas** [comment]
> there will be bug report for this down the line for long running processes

so, I added a public method to explicitly reset the memoized version, for the long running process use case. This looks much more reasonable to me than fetching the key on every fect, as even multi-fetch are not always grouped (eg RedisCluster or RedisArray), thus would pay an extra cost at each fetch.

This require long running processes will then be able to do something like:
`$pool->enableVersioning() || $pool->enableVersioning(false);`

Yes, this is a "new feature" (because a new public method is). Yet, it is required to compensate for a behavior change (long running process don't have their cache cleared anymore). The behavior change is also technically a BC break. Yet also I think this is a really edge-case one and the issue is at the design level so that it cannot be solved otherwise.

**Comment by jderusse** [comment]
A use case for such each case would be to implement a reverse proxy base on ReactPHP (php-pm for instance). With cache invalidation with the clear method.
(I'm not sur it exists and it'd a good idea.. just a brainstorm idea of what could be wrong)

**Comment by andrerom** [comment]
Yes, ReactPHP or similar projects will hit this. But I agree with @nicolas-grekas that other use cases should not having to have a penalty because of it. One option would be to clearly document handlers/ connection classes that don't support flushing pool and hence not recommended for use with long running php processes such as ReactPHP/cron/script..

**Comment by nicolas-grekas** [comment]
> to clearly document handlers/ connection classes that don't support flushing

well, with this patch, *all* handlers do support flushing. But for long running processes, the concern leaks, and I have no better reasonable idea to make it not.

So, we would require ppl to call `$pool->enableVersioning()` on pools that have versioning enabled.
(or call `$pool->enableVersioning() || $pool->enableVersioning(false)`, which works whether versioning is on or not, without changing the versioning-enabled state.)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

File: src/Symfony/Component/Cache/Traits/AbstractTrait.php [hunk 2]
```diff
@@ -159,6 +169,27 @@ public function deleteItems(array $keys)
     }
 
     /**
+     * Enables/disables versioning of items.
+     *
+     * When versioning is enabled, clearing the cache is atomic and doesn't require listing existing keys to proceed,
+     * but old keys may need garbage collection and extra round-trips to the back-end are required.
+     *
+     * Calling this method also clears the memoized namespace version and thus forces a resynchonization of it.
```
Comment (robfrawley) [inline src/Symfony/Component/Cache/Traits/AbstractTrait.php]: @nicolas-grekas Spelling error "memoized" -> "memorized"

File: src/Symfony/Component/Cache/Traits/MemcachedTrait.php [hunk 3]
```diff
@@ -242,7 +243,7 @@ protected function doDelete(array $ids)
      */
     protected function doClear($namespace)
     {
-        return $this->checkResultCode($this->getClient()->flush());
+        return false;
```
Comment (robfrawley) [inline src/Symfony/Component/Cache/Traits/MemcachedTrait.php]: @nicolas-grekas Why has the ability to flush the memcached items been removed?!

File: src/Symfony/Component/Cache/Traits/AbstractTrait.php (hunk 2 above)
Comment (javiereguiluz) [reply src/Symfony/Component/Cache/Traits/AbstractTrait.php]: I reported the "memoized" typo once ... and I was wrong :(  "Memoize" seems to be a valid concept: https://en.wikipedia.org/wiki/Memoization  Maybe it's used correctly this time too?

Comment (robfrawley) [reply src/Symfony/Component/Cache/Traits/AbstractTrait.php]: Ah; seems you are correct! I think the latter one is a spelling error though. ;-)

Comment (nicolas-grekas) [reply src/Symfony/Component/Cache/Traits/AbstractTrait.php]: I wrote memoize on purpose :)

**Pull request merged**